pkg/webhook/stormservice_webhook.go

```diff
@@ -0,0 +1,244 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	orchestrationv1alpha1 "github.com/vllm-project/aibrix/api/orchestration/v1alpha1"
+)
+
+// SetupStormServiceWebhookWithManager registers the webhook for StormService in the manager.
+func SetupStormServiceWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).For(&orchestrationv1alpha1.StormService{}).
+		WithValidator(&StormServiceCustomDefaulter{}).
+		WithDefaulter(&StormServiceCustomDefaulter{}).
+		Complete()
+}
+
+type StormServiceCustomDefaulter struct {
+}
+
+//+kubebuilder:webhook:path=/mutate-orchestration-aibrix-ai-v1alpha1-stormservice,mutating=true,failurePolicy=fail,sideEffects=None,groups=orchestration.aibrix.ai,resources=stormservices,verbs=create;update,versions=v1alpha1,name=mstormservice.kb.io,admissionReviewVersions=v1
+
+var _ webhook.CustomDefaulter = &StormServiceCustomDefaulter{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *StormServiceCustomDefaulter) Default(_ context.Context, obj runtime.Object) error {
+	stormService, ok := obj.(*orchestrationv1alpha1.StormService)
+	if !ok {
+		return fmt.Errorf("expected a StormService object but got %T", obj)
+	}
+
+	// Only proceed if the sidecar injection annotation is present
+	if _, exists := stormService.GetAnnotations()[SidecarInjectionAnnotation]; !exists {
+		return nil
+	}
+
+	// Skip if spec is nil
+	if stormService.Spec.Template.Spec == nil {
+		return nil
+	}
+
+	// Inject sidecar into each role
+	r.injectAIBrixRuntime(stormService)
+
+	return nil
+}
+
+// injectAIBrixRuntime injects the aibrix-runtime sidecar into each Role's pod template
+func (r *StormServiceCustomDefaulter) injectAIBrixRuntime(stormService *orchestrationv1alpha1.StormService) {
+	spec := stormService.Spec.Template.Spec
+
+	for i := range spec.Roles {
+		role := &spec.Roles[i]
+
+		// Skip if sidecar already exists
+		if containsContainer(role.Template.Spec.Containers, SidecarName) {
+			continue
+		}
+
+		// Infer inference engine from primary containers
+		engineType := r.inferEngineType(role.Template.Spec.Containers)
```

Comment:
search `model.aibrix.ai/engine` usage, we can use this value to check the engine type. If we can not find it, then we fallback to your `infer` logic